Split contributing docs to multiple files (#36969)
mistercrunch/Airflow: contributing-docs/testing/unit_tests.rst

* ``BranchExternalPython`` - tests for BranchExternalPythonOperator - selected directly as TestBranchExternalPythonOperator

We have also tests that run "all" tests (so they do not look at the folder, but at the ``pytest`` markers
the tests are marked with to run with some filters applied.

* ``All-Postgres`` - tests that require Postgres database. They are only run when backend is Postgres (``backend("postgres")`` marker)
* ``All-MySQL`` - tests that require MySQL database. They are only run when backend is MySQL (``backend("mysql")`` marker)
* ``All-Quarantined`` - tests that are flaky and need to be fixed (``quarantined`` marker)
* ``All`` - all tests are run (this is the default)

We also have ``Integration`` tests that are running Integration tests with external software that is run
via ``--integration`` flag in ``breeze`` environment - via ``breeze testing integration-tests``.

* ``Integration`` - tests that require external integration images running in docker-compose

This is done for three reasons:

1. in order to selectively run only subset of the test types for some PRs
2. in order to allow efficient parallel test execution of the tests on Self-Hosted runners

> For case 2. We can utilise memory and CPUs available on both CI and local development machines to run

Reviewer: British or US?
```suggestion
For case 2. We can utilize memory and CPUs available on both CI and local development machines to run
```

databases while they are executing.

DB and non-DB tests
-------------------

There are two kinds of unit tests in Airflow - DB and non-DB tests.

Some of the tests of Airflow (around 7000 of them on October 2023)
require a database to connect to in order to run. Those tests store and read data from Airflow DB using
Airflow's core code and it's crucial to run the tests against all real databases that Airflow supports in order
to check if the SQLAlchemy queries are correct and if the database schema is correct.

Those tests should be marked with ``@pytest.mark.db`` decorator on one of the levels:

* test method can be marked with ``@pytest.mark.db`` decorator
* test class can be marked with ``@pytest.mark.db`` decorator
* test module can be marked with ``pytestmark = pytest.mark.db`` at the top level of the module

Airflow's CI runs different test kinds separately.

For the DB tests, they are run against the multiple databases Airflow support, multiple versions of those
and multiple Python versions it supports. In order to save time for testing not all combinations are
tested but enough various combinations are tested to detect potential problems.

> As of October 2023, Airflow has ~9000 Non-DB tests and around 7000 DB tests.
> 
> Airflow non-DB tests
> ....................
> 

Reviewer: Cound of DB tests is mentioned above already, would not repeat this here. Also would move it _into_ the chapter.
```suggestion
Airflow non-DB tests
....................

As of October 2023, Airflow has ~9000 Non-DB tests.

```

Author: Removed the numbers altogether. It adds no value.

For the Non-DB tests, they are run once for each tested Python version with ``none`` database backend (which
causes any database access to fail. Those tests are run with ``pytest-xdist`` plugin in parallel which
means that we can efficiently utilised multi-processor machines (including ``self-hosted`` runners with
8 CPUS we have to run the tests with maximum parallelism).

It's usually straightforward to run those tests in local virtualenv because they do not require any
setup or running database. They also run much faster than DB tests. You can run them with ``pytest`` command
or with ``breeze`` that has all the dependencies needed to run all tests automatically installed. Of course
you can also select just specific test or folder or module for the Pytest to collect/run tests from there,

> the example below shows how to run all tests, parallelising them with ``pytest-xdist``

Reviewer: British or US?
```suggestion
the example below shows how to run all tests, parallelizing them with ``pytest-xdist``
```

    breeze testing non-db-tests

You can pass ``--parallel-test-type`` list of test types to execute or ``--exclude--parallel-test-types``
to exclude them from the default set:.

.. code-block:: bash

    breeze testing non-db-tests --parallel-test-types "Providers API CLI"

.. code-block:: bash

    breeze testing non-db-tests --exclude-parallel-test-types "Providers API CLI"

You can also run the same commands via ``breeze testing tests`` - by adding the necessary flags manually:

.. code-block:: bash

    breeze testing tests --skip-db-tests --backend none --use-xdist

Also you can enter interactive shell with ``breeze`` and run tests from there if you want to iterate
with the tests. Source files in ``breeze`` are mounted as volumes so you can modify them locally and
rerun in Breeze as you will (``-n auto`` will parallelize tests using ``pytest-xdist`` plugin):

.. code-block:: bash

    breeze shell --backend none --python 3.8
    > pytest tests --skip-db-tests -n auto

Airflow DB tests
................

Airflow DB tests require database to run. It can be any of the supported Airflow Databases and they can
be run either using local virtualenv or Breeze

> 

Reviewer: DB tests are mentioned above as well - is this a redundancy-leftover or a history merge conflict?

Author: Yeah. I  rearranged it and moved things together and removed some duplication here.

DB tests (as happens in our CI for the ``Database`` runs) you can use ``--run-db-tests-only`` flag to filter
out non-DB tests (and obviously you can specify not only on the whole ``tests`` directory but on any
folders/files/tests selection, ``pytest`` supports).

.. code-block:: bash

    pytest tests/ --run-db-tests-only

You can also run DB tests with ``breeze`` dockerized environment. You can choose backend to use with
``--backend`` flag. The default is ``sqlite`` but you can also use others such as ``postgres`` or ``mysql``.
You can also select backend version and Python version to use. You can specify the ``test-type`` to run -
breeze will list the test types you can run with ``--help`` and provide auto-complete for them. Example
below runs the ``Core`` tests with ``postgres`` backend and ``3.8`` Python version:

We have a dedicated, opinionated ``breeze testing db-tests`` command as well that runs DB tests
(it is also used in CI to run the DB tests, where you do not have to specify extra flags for
parallel running and you can run all the DB tests
(or just a subset of them with ``--parallel-test-types`` or ``--exclude-parallel-test-types``) in parallel:

.. code-block:: bash

>     breeze testing non-db-tests --backent postgres

Reviewer: Typo?
```suggestion
    breeze testing non-db-tests --backend postgres
```

You should strive to write "pure" non-db unit tests (i.e. DB tests) but sometimes it's just better to plug-in
the existing framework of DagRuns, Dags, Connections and Variables to use the Database directly rather
than having to mock the DB access for example. It's up to you to decide.

However, if you choose to write DB tests you have to make sure you add the ``db_test`` marker - either to
the test method, class (with decorator) or whole module (with pytestmark at the top level of the module).

In most cases when you add tests to existing modules or classes, you follow similar tests so you do not
have to do anything, but in some cases you need to decide if your test should be marked as DB test or
whether it should be changed to not use the database at all.

If your test accesses the database but is not marked properly the Non-DB test in CI will fail with this message:

.. code ::

    "Your test accessed the DB but `_AIRFLOW_SKIP_DB_TESTS` is set.
    Either make sure your test does not use database or mark your test with `@pytest.mark.db_test`.

Marking test as DB test
.......................

> 

Reviewer: Here as well, this chapter smells redundant to the one above?

Author: I moved it above, and renamed it to:

```
Examples of marking test as DB test
```

Code Coverage
-------------

Airflow's CI process automatically uploads the code coverage report to codecov.io.
For the most recent coverage report of the main branch, visit: https://codecov.io/gh/apache/airflow.

Generating Local Coverage Reports:
..................................

If you wish to obtain coverage reports for specific areas of the codebase on your local machine, follow these steps:

a. Initiate a breeze shell.

b. Execute one of the commands below based on the desired coverage area:

   - **Core:** ``python scripts/cov/core_coverage.py``
   - **REST API:** ``python scripts/cov/restapi_coverage.py``
   - **CLI:** ``python scripts/cov/cli_coverage.py``
   - **Webserver:** ``python scripts/cov/www_coverage.py``

c. After execution, the coverage report will be available at: http://localhost:28000/dev/coverage/index.html.

 .. note::

     In order to see the coverage report, you must start webserver first in breeze environment via `airflow webserver`.
     Once you enter `breeze`, you can start `tmux`  (terminal multiplexer) and split the terminal (by pressing `ctrl-B "` for example)

>      to contiinue testing and run the webserver in one tetminal and run tests in the second one (you can switch between

Reviewer: ```suggestion
     to continue testing and run the webserver in one terminal and run tests in the second one (you can switch between
```
